### PR TITLE
ASoC: SOF: debug: Introduce debug option to test DSP ops

### PR DESCRIPTION
**Pull request by ranj063**
Introduce a new debug option to allow testing DSP operations such a boot firware, set power state etc. Add new new kconfig to allow this option for developers. For now only 2 ops are supported for booting a new firmware and setting the DSP power state to D3.

To allow passing a new firmware file, modify the signature of the load_firmware op to take the firmware file name as an argument. When the kconfig for DSP op testing is enabled, runtime PM is disabled to allow the DSP state to be solely managed from the userspace and no audio card is registered.

### CODE REVIEW

File: sound/soc/sof/Kconfig [hunk 1]
```diff
@@ -266,6 +266,14 @@ config SND_SOC_SOF_DEBUG_RETAIN_DSP_CONTEXT
 	  Say Y if you want to retain DSP context for FW exceptions.
 	  If unsure, select "N".
 
+config SND_SOC_SOF_DEBUG_DSP_OPS_TEST
+	bool "SOF enable DSP ops testing"
+	help
+	  This option is used for testing DSP operations such as load/unload FW,
+	  set power state, enable/disable core etc. from userspace.
+	  Say Y if you want to retain DSP context for FW exceptions.
```
Comment (plbossart) [inline sound/soc/sof/Kconfig:274]: that line looks like a copy-paste from above?

File: sound/soc/sof/core.c [hunk 2]
```diff
@@ -472,6 +472,7 @@ static int sof_probe_continue(struct snd_sof_dev *sdev)
 	/* hereafter all FW boot flows are for PM reasons */
 	sdev->first_boot = false;
 
+#if !IS_ENABLED(CONFIG_SND_SOC_SOF_DEBUG_DSP_OPS_TEST)
```
Comment (plbossart) [inline sound/soc/sof/core.c:475]: can we use if (!IS_ENABLED) instead of #if

File: sound/soc/sof/debug-dsp-ops.c [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+		sdev->enabled_cores_mask = 0;
+		sof_set_fw_state(sdev, SOF_FW_BOOT_NOT_STARTED);
+	}
+
+	return 0;
+}
+
+/*
+ * test firmware boot by passing the firmware file as the argument:
+ * ex: echo boot_firmware,intel/avs/tgl/community/dsp_basefw.bin > dsp_test_op
```
Comment (plbossart) [inline sound/soc/sof/debug-dsp-ops.c:56]: why combine the two parts?

echo intel/avs/tgl/community/dsp_basefw.bin > filename
echo 1 > boot

File: sound/soc/sof/debug-dsp-ops.c [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+		return ret;
+
+	sof_set_fw_state(sdev, SOF_FW_BOOT_IN_PROGRESS);
+
+	ret = snd_sof_run_firmware(sdev);
+	if (ret < 0)
+		return ret;
+
+	/* resume DMA trace */
+	return sof_fw_trace_resume(sdev);
```
Comment (plbossart) [inline sound/soc/sof/debug-dsp-ops.c:88]: this looks like code duplication, I don't follow what it's supposed to do

Do we really want to do this or run something like the python stuff on top of a simplified SOF driver.

We have to be really careful not to open a pandora box of new stuff that we'll have to maintain forever twice.

File: sound/soc/sof/sof-priv.h [hunk 5]
```diff
@@ -714,8 +714,8 @@ int sof_create_ipc_file_profile(struct snd_sof_dev *sdev,
 /*
  * Firmware loading.
  */
-int snd_sof_load_firmware_raw(struct snd_sof_dev *sdev);
-int snd_sof_load_firmware_memcpy(struct snd_sof_dev *sdev);
+int snd_sof_load_firmware_raw(struct snd_sof_dev *sdev, const char *fw_filename);
+int snd_sof_load_firmware_memcpy(struct snd_sof_dev *sdev, const char *fw_filename);
```
Comment (plbossart) [inline sound/soc/sof/sof-priv.h:718]: can this go to a separate patch?

**Review by plbossart — COMMENTED** [review]
Not following the general direction if I am honest.

File: sound/soc/sof/debug-dsp-ops.c (hunk 4 above)
Comment (ranj063) [reply sound/soc/sof/debug-dsp-ops.c:88]: @plbossart the idea is to be able to run the python scripts on Linux with minimal effort by using the debugfs option. This is in essence a simplified SOF driver isnt it?

Comment (marc-hb) [reply sound/soc/sof/debug-dsp-ops.c:88]: So is this basically an alternative to https://github.com/thesofproject/sof-diagnostic-driver that don't have to load/unload?

Isn't the sof-diagnostic-driver approach more generic/flexible?

Comment (plbossart) [reply sound/soc/sof/debug-dsp-ops.c:88]: Well, the question is really "what is the partitioning between scripts and driver"? Do we want to reuse snd_sof_run_firmware and hard-code it in a sequence used by a script, or do we want the scripts to do something that runs the firmware?
If you really want to use scripts, then the driver part should be rather minimal and probably a pass-through to access registers.

Comment (ranj063) [reply sound/soc/sof/debug-dsp-ops.c:88]: but @plbossart if we want to write scripts to do all the register reads/writes, it would essentially be rewriting the SOF driver. Isn't this a saner approach?

Comment (ranj063) [reply sound/soc/sof/debug-dsp-ops.c:88]: > So is this basically an alternative to https://github.com/thesofproject/sof-diagnostic-driver that don't have to load/unload?
> 
> Isn't the sof-diagnostic-driver approach more generic/flexible?

@marc-hb this isnt a replacement for the diagnostic driver. This is a way to enable us to run the python scripts for FW testing on Linux with minimal effort.

Comment (plbossart) [reply sound/soc/sof/debug-dsp-ops.c:88]: > but @plbossart if we want to write scripts to do all the register reads/writes, it would essentially be rewriting the SOF driver. Isn't this a saner approach?

the question is whether
a) you want to reuse the existing Python scripts used by FW validation or
b) you want to develop NEW python scripts based on a TBD API

It's not clear from this PR what the direction is. 
a) is complicated because I don't know what the interface requirements are, but 
b) doesn't seem very appealing to me since it'd be a 3rd way of interacting with the DSP.

**Comment by marc-hb** [comment]
https://github.com/zephyrproject-rtos/zephyr/ already has a pure-Python driver from @andyross for testing purposes:

https://github.com/zephyrproject-rtos/zephyr/commit/9073db4323d99757762e63da4135e6ff99516155

```
Newer, much simpler firmware load tool

This is a smaller firmware loader with fewer dependencies and much
faster operation:

+ No need for an externally built SOF diag_driver, it gets its DMA
  memory in userspace and works with any unmodified kernel (that has
  hugetlbfs anyway)

+ Does not leak kernel memory on failure (diag_driver was basically a
  front end for kmalloc(), and if the script exited early...)

+ Much smaller: 230 lines of python in one file vs. 1600 in nine.

+ Much faster; no needless operations and sleeping steps.  Completes
  load and launches hello_world in 0.2s of real time.

+ Correctly resets the stream state and can actually recover form the
  "wedged DSP" state that the previous loader would sometimes get
  stuck in.

+ Clearer structure, easier to use as a testbed for driver-side
  interaction.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>
```
This was the first, very old commit, newer version is now here:
https://github.com/zephyrproject-rtos/zephyr/tree/main/soc/xtensa/intel_adsp/tools

File: sound/soc/sof/debug-dsp-ops.c (hunk 4 above)
Comment (marc-hb) [reply sound/soc/sof/debug-dsp-ops.c:88]: > @marc-hb this isnt a replacement for the diagnostic driver. This is a way to enable us to run the python scripts for FW testing on Linux with minimal effort.

What other purpose does the sof-diagnostic-driver have?

**Comment by andyross** [comment]
@marc-hb To be fair: cavstool isn't PM-aware, so you generally have to unload the kernel driver in order to use it.  But it's been very successful as a paradigm.  I even banged out a broad equivalent for bringing up Zephyr on the mt8195 DSP (which has a MUCH simpler host control interface, again to be fair).

I don't personally see any reason not to have this in the kernel, it seems like an obvious thing for the driver to provide.  I will say though that I'd prefer to see it start from the perspective of "loading arbitrary firmware to the device and getting output" instead of just assuming everything is an audio blob that handles ALSA streams.  The SOF unit testing story for DSP code has always been kinda weak, and Zephyr has had to pick up a lot of that, historically.

File: sound/soc/sof/Kconfig (hunk 1 above)
Comment (ujfalusi) [reply sound/soc/sof/Kconfig:274]: I would mention that this will disable the audio support completely

File: sound/soc/sof/debug-dsp-ops.c (hunk 4 above)
Comment (ujfalusi) [reply sound/soc/sof/debug-dsp-ops.c:88]: @ranj063, it depends _which_ python script we are talking about... One set I have seen is implementing IPC message crafting, sending (yep, with register poking and then polling for completion).
It might be hard to get it to download the firmware because of the DMA configuring, that is true.

File: sound/soc/sof/debug-dsp-ops.c [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+}
+
+/* ops are executed as "op_name,argument1,argument2...". For example, to set the DSP power state
+ * to D3: echo "load_firmware,<PATH>/sof-tgl.ri" > dsp_test_op"
```
Comment (ujfalusi) [inline sound/soc/sof/debug-dsp-ops.c:92]: why not separate files for the firmware path, set_power_state, etc?
How one would guess the valid `op_name`s and what they are set at the moment?

File: sound/soc/sof/debug-dsp-ops.c [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+
+	dfse = devm_kzalloc(sdev->dev, sizeof(*dfse), GFP_KERNEL);
+	if (!dfse)
+		return -ENOMEM;
+
+	/* no need to allocate dfse buffer */
+	dfse->type = SOF_DFSENTRY_TYPE_BUF;
+	dfse->sdev = sdev;
+
+	debugfs_create_file("dsp_test_op", 0222, sdev->debugfs_root, dfse, &sof_dsp_ops_tester_fops);
```
Comment (ujfalusi) [inline sound/soc/sof/debug-dsp-ops.c:156]: `dsp_test` or rather separate files under `dsp_test/` directory?

File: sound/soc/sof/loader.c [hunk 8]
```diff
@@ -14,22 +14,23 @@
...
-				plat_data->fw_filename);
-	if (!fw_filename)
-		return -ENOMEM;
+	if (!fw_filename) {
+		fw_filename = kasprintf(GFP_KERNEL, "%s/%s",
+					plat_data->fw_filename_prefix,
+					plat_data->fw_filename);
+		if (!fw_filename)
+			return -ENOMEM;
+	}
```
Comment (ujfalusi) [inline sound/soc/sof/loader.c:33]: Hm, freeing up the passed `fw_filename` is OK?

File: sound/soc/sof/ipc4-loader.c [hunk 9]
```diff
@@ -157,7 +159,7 @@ static size_t sof_ipc4_fw_parse_basefw_ext_man(struct snd_sof_dev *sdev)
 		if (ret)
 			return ret;
 	}
-
+#endif
```
Comment (ujfalusi) [inline sound/soc/sof/ipc4-loader.c:162]: Why are we skipping this part?

File: sound/soc/sof/loader.c [hunk 10]
```diff
@@ -14,22 +14,23 @@
 #include "sof-priv.h"
 #include "ops.h"
 
-int snd_sof_load_firmware_raw(struct snd_sof_dev *sdev)
+int snd_sof_load_firmware_raw(struct snd_sof_dev *sdev, const char *fw_filename)
```
Comment (ujfalusi) [inline sound/soc/sof/loader.c:17]: you should not neglect the `snd_sof_load_firmware_memcpy()`

File: sound/soc/sof/Kconfig [hunk 11]
```diff
@@ -266,6 +266,14 @@ config SND_SOC_SOF_DEBUG_RETAIN_DSP_CONTEXT
 	  Say Y if you want to retain DSP context for FW exceptions.
 	  If unsure, select "N".
 
+config SND_SOC_SOF_DEBUG_DSP_OPS_TEST
+	bool "SOF enable DSP ops testing"
```
Comment (ujfalusi) [inline sound/soc/sof/Kconfig:270]: What does constitute a "DSP ops"? Isn't this something like "DSP test mode"? Should we have a `sof_debug` flag for this to be enabled/disabled instead of re-building the kernel?

File: sound/soc/sof/core.c [hunk 12]
```diff
@@ -488,6 +489,7 @@ static int sof_probe_continue(struct snd_sof_dev *sdev)
 			"error: failed to register machine driver %d\n", ret);
 		goto fw_trace_err;
 	}
+#endif
 
 	ret = sof_register_clients(sdev);
```
Comment (ujfalusi) [inline sound/soc/sof/core.c:494]: Why are we loading clients in DSP test mode? They might interfere with the testing, no?

File: sound/soc/sof/core.c [hunk 13]
```diff
@@ -511,8 +513,11 @@ static int sof_probe_continue(struct snd_sof_dev *sdev)
 	return 0;
 
 sof_machine_err:
+#if !IS_ENABLED(CONFIG_SND_SOC_SOF_DEBUG_DSP_OPS_TEST)
 	snd_sof_machine_unregister(sdev, plat_data);
+
 fw_trace_err:
+#endif
 	sof_fw_trace_free(sdev);
```
Comment (ujfalusi) [inline sound/soc/sof/core.c:521]: We also enable ftrace? That will send messages, again, might interfere with testing?

**Review by ujfalusi — COMMENTED** [review]
@ranj063, imho if this test mode is activated then we should not boot the 'default' firmware on module load but wait for the firmware path to be loaded and the instruction to boot the DSP with it.

If this mode is selected we should be printing a warning that only firmware testing is allowed via some tbd interface and the kconfig should be more 'alarming' to deter users from selecting it. Currently it is 'why not, it should not do much harm' category.

Note: We sort of have an interface to do simple things via the message injector with the default firmware, we need to prevent rpm to keep the DSP up, but we can send arbitrary messages all while we have the sound card..

File: sound/soc/sof/debug-dsp-ops.c (hunk 4 above)
Comment (lgirdwood) [reply sound/soc/sof/debug-dsp-ops.c:88]: @plbossart @ujfalusi @andyross @marc-hb this PR is NOT about IPC ABI validation but a feature to support debug and compliance/stress testing of kernel + SOF + Zephyr functional IP programming flows with userspace scripting (not userspace IP programming). 

We need to test the kernel functional IP code alongside the associated Zephyr IP code together (with the same timing and sequencing) which is not possible using Python to flip bits via mmap(). We already have a Python mmap() CI tool today that verifies IPC ABI compliance only. i.e. not useful to certify IP programming compliance. 

Yes we can do a lot of test with aplay/amixer, but these tools are end user applications not designed to strongly validate/stress the kernel + SOF + Zephyr combination.

e.g. we need to be able to stress (yes, low hanging fruit will come first)
1) D3 -> D0 and vice versa
2) DSP core off/on 
3) S0ix entry/exit
4) FW boot
5) IMR save/restore
6) topology load/unload

Comment (plbossart) [reply sound/soc/sof/debug-dsp-ops.c:88]: I will assert that you can do all of the above with sof-test. Just run thousands of cycles of suspend-resume, module load/unload, etc.

**Comment by marc-hb** [comment]
The ability to make the DSP panic after some time parameter would also be useful to test recovery features like these:
- #1675
- #4780

**Comment by ujfalusi** [comment]
> The ability to make the DSP panic after some time parameter would also be useful to test recovery features like these:
> 
>     * [Firmware recovery #1675](https://github.com/thesofproject/linux/issues/1675)
> 
>     * [ASoC: SOF: ipc4-pcm: Workaround for crashed firmware on system suspend #4780](https://github.com/thesofproject/linux/pull/4780)

It is not a scope of this PR.

**Comment by marc-hb** [comment]
> > The ability to make the DSP panic

>  It is not a scope of this PR.

Understood but would it be a useful stepping stone? Looks like one.

**Comment by ujfalusi** [comment]
> > > The ability to make the DSP panic
> 
> > It is not a scope of this PR.
> 
> Understood but would it be a useful stepping stone? Looks like one.

To be honest I would rather not do such a thing.

**Comment by plbossart** [comment]
I'll close this for now since it's not an active development. We can always re-open this later.
